accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,187 @@
+
+### Should all inbox source generators be tied to a shared framework?
+
+No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.
+
+### Should only source generators be shipped this way?  What about analyzers and code-fixes?
+
+Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it is desirable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.  We should only do this if we have a good reason, stable analyzers that are lanaguage extensions or tied to stable public API should ship non-side-by-side with the compiler as they do today.  Analyzers that need to validate source generator usage, or analyzers that are closely tied to runtime functionality may consider shipping side-by-side with the framework.
+
+### Why in the ref pack, why not some other "analyzer" pack?
+
+The SDK already has a means for acquisition and selection of ref packs.  The inputs to the selection of ref packs are the same as that of source generators (Version, FrameworkReference, and TargetFramework), and the outputs are consumed by the same phase (compile) with the same level of cross-targeting granularity (TargetFramework and references).
+
+It may be desirable to create a separate package just for analyzers to reduce risk in servicing.  We support servicing ref packs today, but we must be careful not to expose new API nor change any assembly identities.  In my opinion it makes more sense to invest efforts in ensuring safe servicing of ref packs, with potential increased complexity in our own builds.  There is benefit in having servicable ref packs regardless and this keeps additional complexity (resolution of additional pack) out of user builds.
+
+### Why in the ref pack, why not runtime pack?
+
+Runtime packs are currently independent of compile.  Sourcing inputs for compile from a runtime-specific package would break this and require significant changes to the user experience, such as permitting cross-compiling by runtime or "targeting" a runtime.
+
+### Why in the ref pack, why not a dedicated NuGet package?
+
+It is a goal for source-generators to be part of the default experience.  Additionally, the default experience should not require a restore step.  As such analyzers cannot be delivered by NuGet packages alone.
```

Comment:
This is a good question.  I wasn't considering it within the scope of this document, but perhaps it should be.  I don't want to dig into details around how to make generators performant, that's more roslyn's territory.

We'll want to cross-gen them in the SDK installer like other design time code.  They need not be cross-gen'ed in the ref-pack nupkg since that is arch agnostic.

@chsienki do you foresee any other requirements around distribution that would impact the performance?  Will it be a problem if we have source-generator DLLs that are 1:1 with libraries?  

Should inbox source generators share any common infrastructure, APIs, or state?  Or perhaps if we find that such things are needed we push the functionality down into the roslyn APIs.

Reply:
We don't forsee in meaningful impact from source generators. We have a lot of experience with analyzers that stress the compiler in similar ways and our own experiments with generators don't show them to contribute meaningfully to the performance. In virtually every case where existing generators are replaced with roslyn source generators we are seeing build improvements (not the case here, but worth calling out). 

The biggest issue we have with generator and analyzer performance is bugs in the analyzers / generators themselves. It's not uncommon for us to get 2-3 bugs per VS release that we end up tracking down to misbehaving generators / analyzers. In all cases we work with those teams for a patch and get it quickly released to the community.

Reply:
I was thinking more specifically about a design concern around number of modules.  Suppose we have 5 generators split into 5 modules, vs 1 module for all.  Do you see this as causing issues?  (Even trivial ones like some perf-gate somewhere notices and slows down a VS insertion).

Reply:
Based on offline conversations we don't think this should be an issue.